src/Raven.Server/Documents/Indexes/Index.cs

```diff
@@ -1170,43 +1169,46 @@ public bool DoIndexingWork(IndexingStatsScope stats, CancellationToken cancellat
                                         _mre.Set();
                                 }
                             }
-                        }
-                        finally
-                        {
+
                             if (writeOperation.IsValueCreated)
                             {
                                 using (var indexWriteOperation = writeOperation.Value)
                                 {
                                     indexWriteOperation.Commit(stats);
                                 }
                             }
-                        }
-
-                        _indexStorage.WriteReferences(CurrentIndexingScope.Current, tx);
-                    }
 
-                    using (stats.For(IndexingOperation.Storage.Commit))
-                    {
-                        tx.InnerTransaction.LowLevelTransaction.RetrieveCommitStats(out CommitStats commitStats);
+                            _indexStorage.WriteReferences(CurrentIndexingScope.Current, tx);
+                        }
 
-                        tx.InnerTransaction.LowLevelTransaction.AfterCommitWhenNewReadTransactionsPrevented += () =>
+                        using (stats.For(IndexingOperation.Storage.Commit))
                         {
-                            if (writeOperation.IsValueCreated)
+                            tx.InnerTransaction.LowLevelTransaction.RetrieveCommitStats(out CommitStats commitStats);
+
+                            tx.InnerTransaction.LowLevelTransaction.AfterCommitWhenNewReadTransactionsPrevented += () =>
```

Comment:
This is not a good idea for this here.
At this point, we are actually running under lock, and that means that we can stall all queries that are coming in because they can't open the tx.
The problem is that RecreateSearcher / etc are all pretty expensive operations, and not stuff that we can do under lock